reconcile/terraform_repo.py

```diff
@@ -0,0 +1,304 @@
+    version: Optional[int]
+
+
+class AWSAccount(BaseModel):
+    name: str
+    uid: str
+    secret: AWSAuthSecret
+
+
+class TFRepo(BaseModel):
+    """This is what the integration outputs and is input for the executor to perform terraform operations"""
+
+    name: str
+    repository: str
+    ref: str
+    project_path: str
+    account: AWSAccount
+    delete: Optional[bool]
+
+
+class ActionPlan(BaseModel):
+    dry_run: bool
+    repos: list[TFRepo]
+
+
+def get_repos(query_func: Callable) -> list[TFRepo]:
+    """Return all terraform repos defined in app-interface"""
+    query_results = query(query_func=query_func).repos
+    if query_results:
+        return [gql_to_tf_repo(repo) for repo in query_results]
+    return []
+
+
+def get_existing_state(state: State) -> list[TFRepo]:
+    """Get the existing state of terraform repos from S3"""
+    repo_list: list[TFRepo] = []
+    keys = state.ls()
+    for key in keys:
+        value = state.get(key.lstrip("/"), None)
+        repo = TFRepo.parse_raw(value)
+
+        if repo is not None:
+            repo_list.append(repo)
+
+    return repo_list
+
+
+def map_repos(repos: list[TFRepo]) -> MutableMapping[str, TFRepo]:
+    """Generate keys for each repo to more easily compare"""
+    repo_map: MutableMapping[str, TFRepo] = dict()
+    for repo in repos:
+        # repo names are unique per app interface instance as enforced in the schema
+        repo_map[repo.name] = repo
+
+    return repo_map
+
+
+def check_ref(repo_url: str, ref: str, path: str) -> None:
+    """
+    Checks whether a Git ref is valid
+    and whether config.tf exists in the project path"""
+    instance = queries.get_gitlab_instance()
+    with GitLabApi(
+        instance, settings=queries.get_secret_reader_settings(), project_url=repo_url
+    ) as gl:
+        try:
+            gl.get_commit_sha(ref=ref, repo_url=repo_url)
+            retrieved_path = gl.get_file("{}/config.tf".format(path), ref)
+
+            if not retrieved_path:
+                raise ParameterError(
+                    'No config.tf found in path: "{}" on repo: "{}". Ensure that a config.tf file is present in this path.'.format(
+                        path, repo_url
+                    )
+                )
+        except (KeyError, AttributeError):
+            raise ParameterError(
+                'Invalid ref: "{}" on repo: "{}"'.format(ref, repo_url)
+            )
+
+
+def diff_missing(
+    a: Mapping[str, TFRepo],
+    b: Mapping[str, TFRepo],
+    requireDelete: bool = False,
+) -> MutableMapping[str, TFRepo]:
+    """
+    Returns a mapping of repos that are present in mapping a but missing in mapping b
+    When requireDelete = True, each repo from a is checked to ensure that the delete flag is set"""
+    missing: MutableMapping[str, TFRepo] = dict()
+    for a_key, a_repo in a.items():
+        b_repo = b.get(a_key)
+        if b_repo is None:
+            if requireDelete is True and a_repo.delete is not True:
```

Comment:
`is True` and `is not True` are not necessary, can directly use

```python
if requireDelete and not a_reop.delete
```

[reference](https://peps.python.org/pep-0008/#programming-recommendations):

```
Don’t compare boolean values to True or False using ==:
# Correct:
if greeting:

# Wrong:
if greeting == True:

Worse:
# Wrong:
if greeting is True:
```